### PR TITLE
Add onEvent methods to Template

### PR DESCRIPTION
**Pull request by Slava**
Add `Template#onCreated`, `Template#onRendered` and `Template#onDestroyed` methods. These are better than assigning `Template.rendered`, etc properties as you can add more than 1 of them w/o overriding previously added event handlers.

Ran test-packages for Blaze.

@avital @dgreensp

### CODE REVIEW

File: packages/blaze/template.js [hunk 1]
```diff
@@ -56,6 +62,52 @@ Blaze.isTemplate = function (t) {
   return (t instanceof Blaze.Template);
 };
 
+/**
+ * @name  onCreated
+ * @instance
+ * @memberOf Template
+ * @summary Add a callback when an instance of a template is created.
```
Comment (stubailo) [inline packages/blaze/template.js]: This summary is incorrect - it doesn't add the callback when the template is created. It add a callback that is called when an instance of a template is created. So maybe: "Register a function to be called when an instance of this template is created."

File: packages/blaze/template.js [hunk 2]
```diff
@@ -56,6 +62,52 @@ Blaze.isTemplate = function (t) {
...
+ */
+Template.prototype.onCreated = function (cb) {
+  this._callbacks.created.push(cb);
+};
+
+/**
+ * @name  onRendered
+ * @instance
+ * @memberOf Template
+ * @summary Add a callback when an instance of a template is rendered.
```
Comment (stubailo) [inline packages/blaze/template.js]: "Register a function to be called when an instance of this template is inserted into the DOM."

File: packages/blaze/template.js [hunk 3]
```diff
@@ -56,6 +62,52 @@ Blaze.isTemplate = function (t) {
...
+ */
+Template.prototype.onRendered = function (cb) {
+  this._callbacks.rendered.push(cb);
+};
+
+/**
+ * @name  onDestroyed
+ * @instance
+ * @memberOf Template
+ * @summary Add a callback when an instance of a template is destroyed.
```
Comment (stubailo) [inline packages/blaze/template.js]: "Register a function to be called when an instance of this template is removed from the DOM and destroyed." (I'm not 100% sure, basically we should describe what destroyed means)

File: packages/blaze/template.js [hunk 4]
```diff
@@ -56,6 +62,52 @@ Blaze.isTemplate = function (t) {
...
+ * @param {Function} callback A function to be added as a callback.
+ * @locus Client
+ */
+Template.prototype.onDestroyed = function (cb) {
+  this._callbacks.destroyed.push(cb);
+};
+
+Template.prototype._fireCallbacks = function (template, which) {
+  var self = this;
+  var callbacks = self[which] ? [self[which]] : [];
```
Comment (stubailo) [inline packages/blaze/template.js]: Could use a comment. So we fire the .rendered callback _or_ .onRendered? But not both? Why not?

Comment (Slava) [inline packages/blaze/template.js]: Both. Added a comment.

File: packages/blaze/template.js (hunk 3 above)
Comment (Slava) [inline packages/blaze/template.js]: I think mentioning the magical "destroyed" helps, although Blaze can remove a tree from DOM and then attach it somewhere else.

**Comment by Slava** [comment]
Updated

**Comment by Slava** [comment]
updated docs

File: docs/client/basic/sections/templates.md [hunk 5]
```diff
@@ -210,27 +210,26 @@ event object and template instance. See the [Event Maps section](#eventmaps)
...
+The functions added with this method are called once for every instance of
 *Template.myTemplate* when it is inserted into the page for the first time.
 
-This _rendered_ callback can be used to integrate external libraries that aren't
-familiar with Meteor's automatic view rendering, and need to be initialized
-every time HTML is inserted into the page. Use the
-[`created`](#template_created) and
-[`destroyed`](#template_destroyed) callbacks to perform
-initialization or clean-up on any objects.
+These _rendered_ callbacks can be used to integrate external libraries that
```
Comment (stubailo) [inline docs/client/basic/sections/templates.md]: I think we can delete _rendered_ here

File: docs/client/basic/sections/templates.md [hunk 6]
```diff
@@ -210,27 +210,26 @@ event object and template instance. See the [Event Maps section](#eventmaps)
...
+aren't familiar with Meteor's automatic view rendering, and need to be
+initialized every time HTML is inserted into the page. Use the callbacks added
```
Comment (stubailo) [inline docs/client/basic/sections/templates.md]: I think we can just say "You can also use `onCreated` and `onDestroyed` to perform initialization or clean-up..."

File: docs/client/basic/toc.js [hunk 7]
```diff
@@ -48,7 +48,7 @@ var sections = [
       item("Defining templates in HTML", {id: "defining-templates"}),
       item("Template.<em>name</em>.helpers", {longname: "Template#helpers"}),
       item("Template.<em>name</em>.events", {longname: "Template#events"}),
-      item("Template.<em>name</em>.rendered", {longname: "Template#rendered"}),
+      item("Template.<em>name</em>.onRendered", {longname: "Template#onRendered"}),
```
Comment (stubailo) [inline docs/client/basic/toc.js]: Do we have a backcompat link for this?

File: docs/client/full-api/api/blaze.md [hunk 8]
```diff
@@ -8,11 +8,11 @@ and manipulate "Views," the building blocks of reactive templates.
 
 {{> autoApiBox "Blaze.render"}}
 
-When you render a template, the
-template's [`created`](#template_created) callback is invoked
-immediately, before evaluating the content of the template.
-The [`rendered`](#template_rendered) callback is
-invoked after the View is rendered and inserted into the DOM.
+When you render a template, the template's callbacks added with
```
Comment (stubailo) [inline docs/client/full-api/api/blaze.md]: delete "template's"

File: docs/client/full-api/api/templates.md [hunk 9]
```diff
@@ -45,44 +45,46 @@ To create a helper that can be used in any template, use
...
+evaluated for the first time. Inside a callback, `this` is the new [template
+instance](#template_inst) object. Properties you set on this object will be
+visible from the callbacks added with `onRendered` and `onDestroyed` methods and
+from event handlers.
 
-This callback fires once and is the first callback to fire.  Every
-`created` has a corresponding `destroyed`; that is, if you get a
-`created` callback with a certain template instance object in `this`,
-you will eventually get a `destroyed` callback for the same object.
+These callbacks fire once and are the first group of callbacks to fire. Every
```
Comment (stubailo) [inline docs/client/full-api/api/templates.md]: Let's remove the "corresponding callbacks" sentence

**Comment by Slava** [comment]
Updated the usage in examples and packages

File: docs/client/basic/toc.js (hunk 7 above)
Comment (Slava) [inline docs/client/basic/toc.js]: yes, see in link-redirect.js

Comment (Slava) [inline docs/client/basic/toc.js]: actually, no, there is no backwards-compatible link redirection for basic docs, you are right.

Can we ignore it please? It is not like hundreds of stackoverflow questions link to that paragraph in basic docs and @debergalis approves breaking it.

Comment (stubailo) [inline docs/client/basic/toc.js]: Literally everyone is relying on this answer: http://stackoverflow.com/questions/26976603/meteor-load-javascript-file-but-dont-execute-execute-on-demand

I can't believe you want to break this link

Comment (debergalis) [inline docs/client/basic/toc.js]: We can and should edit SO answers.

Comment (stubailo) [inline docs/client/basic/toc.js]: Haha the previous comment was a joke :] it's the only answer, so we will just edit it.

**Comment by Slava** [comment]
Merged!

**Pull request merged**